engine/execution/state/state.go

```diff
@@ -306,7 +307,28 @@ func (s *state) StateCommitmentByBlockID(ctx context.Context, blockID flow.Ident
 }
 
 func (s *state) ChunkDataPackByChunkID(ctx context.Context, chunkID flow.Identifier) (*flow.ChunkDataPack, error) {
-	return s.chunkDataPacks.ByChunkID(chunkID)
+	storedChunkDataPack, err := s.chunkDataPacks.ByChunkID(chunkID)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve stored chunk data pack: %w", err)
+	}
+
+	chunkDataPack := &flow.ChunkDataPack{
+		ChunkID:    storedChunkDataPack.ChunkID,
+		StartState: storedChunkDataPack.StartState,
+		Proof:      storedChunkDataPack.Proof,
+	}
+
+	if storedChunkDataPack.CollectionID != nil {
+		// non-system chunks have a non-nil collection
+		collection, err := s.GetCollection(*storedChunkDataPack.CollectionID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve collection for chunk data pack: %w", err)
+		}
+
+		chunkDataPack.Collection = collection
+	}
+
+	return chunkDataPack, nil
```

Comment:
`ChunkDataPack` has a `Collection` field (instead of the older version `CollectionID`). However, on the execution node a `ChunkDataPack` is stored as the `badger` internal type of  `StoredChunkDataPack`. The difference between the `ChunkDataPack` and `StoredChunkDataPack` is the former has a `Collection` filed, but the latter only keeps the `CollectionID` and not the actual collection. This is to avoid redundant storage of the same `Collection` on the execution node at two different places, i.e., a collection is stored at `Collections` badger storage, hence, just knowing its identifier, one can look it up. 

One can see the `StoredChunkDataPack` as a cold-version persistent data type that is converted to the `ChunkDataPack` by fetching its `Collection` upon request.